### PR TITLE
`unnecessary_lazy_eval` show suggestions on multiline lint

### LINKED ISSUE #8393: unnecessary_lazy_evaluations missing / wrong hint information
### Summary

The `unnecessary_lazy_evaluations` can produce a help hint which misses the line of the actual error, producing a hint which is not usable (is a missing a line).

I only have this case I ran into but my code senses make this smell like a general issue - is it possible hint lines get cut off past 5 lines? Clippy should ensure that the actually affected line does not get cut off from any hint size limits.

### Reproducer

I tried this code:

```rust
fn parse_rfc850_date(s: &[u8]) -> Result<HttpDate, DateError> {
    // Example: `Sunday, 06-Nov-94 08:49:37 GMT`

    fn week_day<'a>(s: &'a [u8], week_day: u8, name: &'static [u8]) -> Option<(u8, &'a [u8])> {
        if &s[0..name.len()] == name {
            return Some((week_day, &s[name.len()..]));
        }
        None
    }
    let (week_day, s) = week_day(s, 1, b"Monday, ")
        .or_else(|| week_day(s, 2, b"Tuesday, "))
        .or_else(|| week_day(s, 3, b"Wednesday, "))
        .or_else(|| week_day(s, 4, b"Thursday, "))
        .or_else(|| week_day(s, 5, b"Friday, "))
        .or_else(|| week_day(s, 6, b"Saturday, "))
        .or_else(|| week_day(s, 7, b"Sunday, "))
        .ok_or_else(|| DateError::FormatInvalid("rfc850"))?;
}
```

Causes:
```
warning: unnecessary closure used to substitute value for `Option::None`
   --> src/utils/date.rs:130:25
    |
130 |       let (week_day, s) = week_day(s, 1, b"Monday, ")
    |  _________________________^
131 | |         .or_else(|| week_day(s, 2, b"Tuesday, "))
132 | |         .or_else(|| week_day(s, 3, b"Wednesday, "))
133 | |         .or_else(|| week_day(s, 4, b"Thursday, "))
...   |
136 | |         .or_else(|| week_day(s, 7, b"Sunday, "))
137 | |         .ok_or_else(|| DateError::FormatInvalid("rfc850"))?;
    | |__________________________________________________________^
    |
    = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
```

I expected to see this happen:

```
help: use `ok_or` instead
    |
130 ~     let (week_day, s) = week_day(s, 1, b"Monday, ")
131 +         .or_else(|| week_day(s, 2, b"Tuesday, "))
132 +         .or_else(|| week_day(s, 3, b"Wednesday, "))
133 +         .or_else(|| week_day(s, 4, b"Thursday, "))
134 +         .or_else(|| week_day(s, 5, b"Friday, "))
135 +         .or_else(|| week_day(s, 6, b"Saturday, "))
136 +         .ok_or(DateError::FormatInvalid("rfc850"))?;
  ...
```

Instead, this happened:

```
help: use `ok_or` instead
    |
130 ~     let (week_day, s) = week_day(s, 1, b"Monday, ")
131 +         .or_else(|| week_day(s, 2, b"Tuesday, "))
132 +         .or_else(|| week_day(s, 3, b"Wednesday, "))
133 +         .or_else(|| week_day(s, 4, b"Thursday, "))
134 +         .or_else(|| week_day(s, 5, b"Friday, "))
135 +         .or_else(|| week_day(s, 6, b"Saturday, "))
  ...
```

(It misses the last line which is where the lint error actually is, causing a misleading suggestion)

### Version

```text
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: x86_64-unknown-linux-gnu
release: 1.58.1
LLVM version: 13.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

### PR DESCRIPTION
**Pull request by dswij**
Closes #8393 

changelog: [`unnecessary_lazy_eval`] now shows suggestions for long-line lints

### CODE REVIEW

File: clippy_lints/src/methods/unnecessary_lazy_eval.rs [hunk 1]
```diff
@@ -48,20 +48,16 @@ pub(super) fn check<'tcx>(
...
-                    &format!("use `{}` instead", simplify_using),
-                    format!(
-                        "{0}.{1}({2})",
-                        snippet(cx, recv.span, ".."),
-                        simplify_using,
-                        snippet(cx, body_expr.span, ".."),
-                    ),
-                    applicability,
-                );
+                if let hir::ExprKind::MethodCall(_, _, span) = expr.kind {
```
Comment (dswij) [inline clippy_lints/src/methods/unnecessary_lazy_eval.rs:51]: Not the best way, but I'd prefer not to change this

https://github.com/rust-lang/rust-clippy/blob/b83c6323c7d788fe4b6921b000324fb09945e12b/clippy_lints/src/methods/mod.rs#L2102

Comment (giraffate) [reply clippy_lints/src/methods/unnecessary_lazy_eval.rs:51]: Can the comment about it be added here?

**Review by giraffate — COMMENTED** [review]
LGTM with nits

**Comment by giraffate** [comment]
@bors r+

Thanks!

**Pull request merged**